rclcpp/src/rclcpp/executor.cpp

```diff
@@ -232,16 +232,18 @@ Executor::spin_some(std::chrono::nanoseconds max_duration)
     throw std::runtime_error("spin_some() called while already spinning");
   }
   RCLCPP_SCOPE_EXIT(this->spinning.store(false); );
-  // non-blocking call to pre-load all available work
-  wait_for_work(std::chrono::milliseconds::zero());
-  while (spinning.load() && max_duration_not_elapsed()) {
-    AnyExecutable any_exec;
-    if (get_next_ready_executable(any_exec)) {
-      execute_any_executable(any_exec);
-    } else {
-      break;
+  do {
+    // non-blocking call to pre-load all available work
+    wait_for_work(std::chrono::milliseconds::zero());
+    while (spinning.load() && max_duration_not_elapsed()) {
+      AnyExecutable any_exec;
+      if (get_next_ready_executable(any_exec)) {
+        execute_any_executable(any_exec);
+      } else {
+        return;
```

Comment:
@ivanpauno how's this any different? I'd think `spin_some()` should return only once a non-blocking `wait_for_work()` call results in no executables to be served.

Reply:
:joy:
The original version could introduce a busy loop (non-blocking wait with nothing ready tried repeatedly). https://github.com/ros2/rclcpp/pull/1156/commits/0b2f75f80ef720fcbf1dfdda9c1625096ecba28d was a really bad attempt of fixing that.

See https://github.com/ros2/rclcpp/pull/1156/commits/6866819cae28a6297beebd12940989566610c54d for a real fix.

Reply:
:man_facepalming: see https://github.com/ros2/rclcpp/pull/1156/commits/55fbb44d117839b017e89207651815996af7d5bd.
I tested it this time.